examples/pinn/eikonal.py

```diff
@@ -0,0 +1,459 @@
+synthetic model.
+
+The eikonal equation is a first-order nonlinear PDE (partial differential equation). This
+example is aimed at implementing the specific neural-network architecture NES for
+accurate and robust solving the eikonal equation. Similar networks for solving PDEs like
+eikonal are called Physics-Informed Neural Networks (PINNs) which were introduced by
+[Raissi et. al. (2019)](https://doi.org/10.1016/j.jcp.2018.10.045). The main idea of
+PINNs is to mimic PDEs we want to solve so that the trained network approximates the
+solution to a PDE.
+
+Here, we will be constructing a neural-network architecture for solving the following
+two-point modification of the [eikonal
+equation](https://en.wikipedia.org/wiki/Eikonal_equation):
+
+$$\Vert \nabla_r \tau(\textbf{x}_s, \textbf{x}_r) \Vert = \frac{1}{v(\textbf{x}_r)}$$
+
+where $\textbf{x}_s=(x_s, y_s)$ is spatial coordinate of source or start point (for
+simplicity we consider 2D space), $\textbf{x}_r=(x_r, y_r)$ is receiver or end point,
+$\Vert \cdot \Vert$ is Euclidean norm, $\nabla_r=(\partial_{x_r}, \partial_{y_r})$ is
+gradient operator w.r.t. $\textbf{x}_r$, $v(\textbf{x}_r)$ is velocity at point
+$\textbf{x}_r$, $\tau(\textbf{x}_s, \textbf{x}_r)$ is solution to the equation (eikonal).
+The eikonal $\tau(\textbf{x}_s, \textbf{x}_r)$ describes the fastest travel time from a
+point $\textbf{x}_s$ to a point $\textbf{x}_r$, and, as a consequence, it describes the
+shortest path trajectory. An important part of the equation is its boundary condition:
+
+$$ \tau(\textbf{x}_s, \textbf{x}_s)=0 $$
+
+which states that the travel time at a start point is zero. Another important constraint
+is $\tau \geq 0$.
+
+### Application of eikonal
+
+The eikonal is used in many fields such as computer vision, shortest path problems, image
+segmentation, and modelling of wave propagation. Specifically, this equation plays an
+important role in seismology. It approximates the propagation of seismic waves in the
+Earth and can be used to locate earthquake hypocenters.
+
+### Why PINNs?
+Finite methods (finite differences, finite elements, finite volumes) have been developed
+for decades and achieved high accuracy and speed. However, they still have some
+limitations and in complex settings their application may be complicated. For example, it
+is tricky to apply finite-difference methods when the mesh is irregular in space. In this
+case, an advantage of PINNs is that they are mesh-free, the same neural network can be
+used for any geometry and distribution of collocation points where we seek for solution.
+Another important advantage, it can provide dramatic compression of the solution in
+massive computations. The latter will be shown in the end of the example.
+"""
+
+"""
+## Imports
+"""
+
+import tensorflow as tf
+from tensorflow.keras.layers.experimental.preprocessing import Rescaling
+from tensorflow.keras.layers import Input, Concatenate, Dense
```

Comment:
Just `from tensorflow.keras import layers`, then access elements of the module, e.g. `layers.Dense`